### PR TITLE
CLI improvements - followup

### PR DESCRIPTION
**Pull request by juan518munoz**
# CLI improvements - Followup

## Description

<p align="center">
image
</p>

This is a followup of the [previous cli improvements pr](https://github.com/lambdaclass/lambdaworks/pull/598), it adds to the CLI the following commands that were used to be ran from the `makefile`:

- `compile`
- `compile-and-prove`
- `compile-and-run-all`

The new functionality added makes use of [Rust's Command](https://doc.rust-lang.org/std/process/struct.Command.html) to spawn processes of either `cairo-compile` or `docker run` with the image built from the dockerfile.

## Type of change

- [x] Optimization

## Checklist
- [x] This change requires new documentation.
  - [x] Documentation has been added/updated.

### CODE REVIEW

**Review by diegokingston — APPROVED** [review]

File: provers/cairo-prover-cli/src/main.rs [hunk 1]
```diff
@@ -5,42 +5,107 @@ use lambdaworks_math::field::fields::fft_friendly::stark_252_prime_field::Stark2
...
-}
+/// Attemps to compile the Cairo program
+/// with either `cairo-compile` or `docker``
+fn try_compile(program_path: &String, out_file_path: &String) -> Result<(), Error> {
+    if !program_path.contains(".cairo") {
+        println!("The program provided is not an uncompiled .cairo program");
+        return Err(Error::new(
+            ErrorKind::Other,
+            "File provided is not a Cairo uncompiled",
+        ));
```
Comment (MauroToscano) [inline provers/cairo-prover-cli/src/main.rs:94]: The print should be in the caller that handles the error

File: provers/cairo-prover-cli/src/main.rs [hunk 2]
```diff
@@ -5,42 +5,107 @@ use lambdaworks_math::field::fields::fft_friendly::stark_252_prime_field::Stark2
...
-#[derive(Args, Debug)]
-struct ProveAndVerifyArgs {
-    program_path: String,
+    if cairo_compile(program_path, out_file_path).is_ok()
+        || docker_compile(program_path, out_file_path).is_ok()
+    {
+        println!("Compiled cairo program");
+        Ok(())
+    } else {
+        println!("Failed to compile cairo program\nEnsure cairo-compile or docker is installed (and running)");
```
Comment (MauroToscano) [inline provers/cairo-prover-cli/src/main.rs:103]:
```suggestion
        println!("Failed to compile cairo program\nEnsure either cairo-compile is installed, or Docker is running and the compiler image is built");
```

**Review by MauroToscano — CHANGES_REQUESTED** [review]

File: provers/cairo-prover-cli/src/main.rs (hunk 2 above)
Comment (juan518munoz) [reply provers/cairo-prover-cli/src/main.rs]: Removed print altogether and modified `Error` return after highlighted code:

```rs
        Err(Error::new(
            ErrorKind::Other,
            "Failed to compile cairo program, neither cairo-compile nor docker found",
        ))
```

Which is then printed out by the caller.

File: provers/cairo-prover-cli/src/main.rs (hunk 1 above)
Comment (juan518munoz) [reply provers/cairo-prover-cli/src/main.rs]: Applied suggestion.

File: provers/cairo-prover-cli/src/main.rs [hunk 3]
```diff
@@ -5,42 +5,104 @@ use lambdaworks_math::field::fields::fft_friendly::stark_252_prime_field::Stark2
...
-#[derive(Args, Debug)]
-struct ProveAndVerifyArgs {
-    program_path: String,
+/// Attemps to compile the Cairo program
+/// with either `cairo-compile` or `docker``
+fn try_compile(program_path: &String, out_file_path: &String) -> Result<(), Error> {
+    if !program_path.contains(".cairo") {
+        return Err(Error::new(
+            ErrorKind::Other,
+            "File provided is not a Cairo uncompiled",
```
Comment (entropidelic) [inline provers/cairo-prover-cli/src/main.rs:92]:
```suggestion
            "Provided file is not a Cairo source file",
```

**Review by entropidelic — APPROVED** [review]

**Comment by entropidelic** [comment]
nice one 🚀

**Comment by entropidelic** [comment]
Let's wait to merge this one until #596 is merged, so that no further conflicts appear on that PR.

File: provers/cairo/src/main.rs [hunk 4]
```diff
@@ -158,12 +228,38 @@ fn main() {
...
             verify_proof(proof, pub_inputs, &proof_options);
         }
+        commands::ProverEntity::CompileAndProve(args) => {
+            let out_file_path = args.program_path.replace(".cairo", ".json");
+            match try_compile(&args.program_path, &out_file_path) {
+                Ok(_) => {
+                    generate_proof(&out_file_path, &proof_options);
+                }
+                Err(err) => {
+                    println!("{}", err)
```
Comment (entropidelic) [inline provers/cairo/src/main.rs:245]: small nit, `eprintln!` can be used to print to standard error

File: provers/cairo/src/main.rs [hunk 5]
```diff
@@ -158,12 +228,38 @@ fn main() {
...
+            match try_compile(&args.program_path, &out_file_path) {
+                Ok(_) => {
+                    let Some((proof, pub_inputs)) = generate_proof(&out_file_path, &proof_options)
+                    else {
+                        return;
+                    };
+                    verify_proof(proof, pub_inputs, &proof_options);
+                }
+                Err(err) => {
+                    println!("{}", err)
```
Comment (entropidelic) [inline provers/cairo/src/main.rs:260]: same here, sorry I missed this in the previous review

**Review by juan518munoz — COMMENTED** [review]
Good catch! applied suggested changes and also modifies this `error` print inside `generate_proof`:

https://github.com/lambdaclass/lambdaworks/blob/ed579508746189a81fb4715cdec6478202fdac99/provers/cairo/src/main.rs#L133-L136

File: provers/cairo/src/main.rs [hunk 6]
```diff
@@ -130,8 +130,8 @@ fn generate_proof(
     println!("Making proof ...");
     let proof = match generate_cairo_proof(&main_trace, &pub_inputs, proof_options) {
         Ok(p) => p,
-        Err(e) => {
-            println!("Error generating proof: {:?}", e);
+        Err(err) => {
+            eprintln!("Error generating proof: {:?}", err);
```
Comment (juan518munoz) [inline provers/cairo/src/main.rs:134]: forgot to quote code

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**